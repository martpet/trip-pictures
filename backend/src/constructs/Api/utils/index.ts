export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
