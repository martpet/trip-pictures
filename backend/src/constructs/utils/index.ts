export * from './createEdgeFunction';
export * from './createNodejsFunction';
export * from './createTable';
