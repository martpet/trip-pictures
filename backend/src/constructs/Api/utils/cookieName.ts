import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
