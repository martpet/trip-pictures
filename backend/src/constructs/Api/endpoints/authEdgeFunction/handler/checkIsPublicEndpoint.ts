import { isPublicEndpoint } from '~/../../shared/utils/isPublicEndpoint';
import { LambdaEdgeViewerEvent } from '~/constructs/Api/types';

export function checkIsPublicEndpoint(edgeEvent: LambdaEdgeViewerEvent) {
  const { uri, method } = edgeEvent.Records[0].cf.request;

  return isPublicEndpoint({
    path: uri,
    method,
  });
}
