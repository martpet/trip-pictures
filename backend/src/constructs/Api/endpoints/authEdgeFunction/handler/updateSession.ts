import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, UpdateCommand } from '@aws-sdk/lib-dynamodb';

import { region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';
import { createDynamoUpdateExpression } from '~/utils';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

type UpdateSessionProps = {
  sessionId: string;
  idToken: string;
};

export const updateSession = ({ sessionId, idToken }: UpdateSessionProps) => {
  const { tableName, partitionKey } = sessionsTableOptions;

  const data: Partial<SessionsTableItem> = {
    idToken,
  };

  const updateCommand = new UpdateCommand({
    TableName: tableName,
    Key: { [partitionKey.name]: sessionId },
    ...createDynamoUpdateExpression(data),
  });

  return ddbDocClient.send(updateCommand);
};
