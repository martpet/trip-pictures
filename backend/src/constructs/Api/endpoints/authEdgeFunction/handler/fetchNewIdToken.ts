import fetch from 'node-fetch';

import { FetchNewIdTokenResponse } from '~/constructs/Api/types';
import { authPaths } from '~/consts';

import { updateSession } from './updateSession';

type FetchNewIdTokenProps = {
  refreshToken: string;
  sessionId: string;
  clientId: string;
};

export const fetchNewIdToken = async ({
  refreshToken,
  sessionId,
  clientId,
}: FetchNewIdTokenProps) => {
  const { authDomain } = globalAuthEdgeFunctionProps;

  console.info('Fetching new id token...');

  const response = await fetch(`https://${authDomain}${authPaths.token}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'refresh_token',
      client_id: clientId,
      refresh_token: refreshToken,
    }),
  });

  const data = (await response.json()) as FetchNewIdTokenResponse;

  if ('error' in data) {
    throw new Error(data.error);
  }

  await updateSession({ sessionId, idToken: data.id_token });

  console.info('New id token fetched and saved');

  return data.id_token;
};
