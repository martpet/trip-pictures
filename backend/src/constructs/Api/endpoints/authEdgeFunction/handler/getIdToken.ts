import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';

import { getIdTokenPayload } from '~/constructs/Api/utils/getIdTokenPayload';
import { refreshTokenExpiredErrorMessage, region, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

import { fetchNewIdToken } from './fetchNewIdToken';

const ddbClient = new DynamoDBClient({ region });
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

export const getIdToken = async (sessionId: string) => {
  const getCommand = new GetCommand({
    TableName: sessionsTableOptions.tableName,
    Key: { [sessionsTableOptions.partitionKey.name]: sessionId },
  });

  const { Item } = await ddbDocClient.send(getCommand);
  const sessionsItem = Item as SessionsTableItem | undefined;

  if (!sessionsItem) {
    throw new Error(`could not find session with id "${sessionId}"`);
  }

  const { idToken, refreshToken, refreshTokenExpires } = sessionsItem;
  const { exp: idTokenExpires, aud: clientId } = await getIdTokenPayload(idToken);

  if ((idTokenExpires - 5) * 1000 > Date.now()) {
    return idToken;
  }

  if (refreshTokenExpires * 1000 > Date.now()) {
    return fetchNewIdToken({ refreshToken, sessionId, clientId });
  }

  throw new Error(refreshTokenExpiredErrorMessage);
};
