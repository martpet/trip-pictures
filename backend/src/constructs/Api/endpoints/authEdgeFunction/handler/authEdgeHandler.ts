import clone from 'clone';
import { StatusCodes } from 'http-status-codes';

import { LambdaEdgeViewerRequestHandler } from '~/constructs/Api/types';
import { edgeErrorResponse } from '~/constructs/Api/utils/edgeErrorResponse';
import { parseLambdaEdgeEventCookies } from '~/constructs/Api/utils/parseLambdaEdgeEventCookies';

import { checkIsPublicEndpoint } from './checkIsPublicEndpoint';
import { getIdToken } from './getIdToken';

export const handler: LambdaEdgeViewerRequestHandler = async (event) => {
  try {
    const { request } = event.Records[0].cf;
    delete request.headers.authorization;

    if (request.method === 'OPTIONS') {
      return { status: StatusCodes.NO_CONTENT };
    }

    if (checkIsPublicEndpoint(event)) {
      return request;
    }

    const { sessionId } = parseLambdaEdgeEventCookies(event);

    if (!sessionId) {
      throw new Error('invalid or missing `sessionId` cookie');
    }

    const requestClone = clone(request);
    requestClone.headers.authorization = [{ value: await getIdToken(sessionId) }];
    return requestClone;
  } catch (error) {
    return edgeErrorResponse('eWUjcKPBi3', {
      statusCode: StatusCodes.UNAUTHORIZED,
      error,
    });
  }
};
