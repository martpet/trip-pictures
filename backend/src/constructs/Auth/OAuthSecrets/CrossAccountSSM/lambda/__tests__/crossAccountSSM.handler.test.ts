import { GetParametersCommand, SSMClient } from '@aws-sdk/client-ssm';
import { mockClient } from 'aws-sdk-client-mock';

import {
  itCalls,
  itRejects,
  itResolves,
  itSendsAwsCommand,
} from '~/constructs/Api/utils';
import { getRoleCredentials } from '~/utils';

import { handler } from '../crossAccountSSM.handler';

vi.mock('~/utils/getRoleCredentials');

const ssmMock = mockClient(SSMClient);

const args = [
  {
    ResourceProperties: {
      roleArn: 'dummyRoleArn',
      getParametersInput: {
        Names: ['dummyParameterName1', 'dummyParameterName2'],
      },
    },
  },
] as unknown as Parameters<typeof handler>;

beforeEach(() => {
  ssmMock.reset();

  ssmMock.on(GetParametersCommand).resolves({
    Parameters: [
      {
        Name: 'dummyParameterName2',
        Value: 'dummyParameterValue2',
      },
      {
        Name: 'dummyParameterName1',
        Value: 'dummyParameterValue1',
      },
    ],
  });
});

describe('crossAccountSSM.handler', () => {
  itCalls(getRoleCredentials, handler, args);
  itSendsAwsCommand(GetParametersCommand, ssmMock, handler, args);
  itResolves(handler, args);

  describe('when "getParametersInput.Names" is missing', () => {
    const argsClone = structuredClone(args);
    delete argsClone[0].ResourceProperties.getParametersInput.Names;
    itRejects(handler, argsClone);
  });

  describe('when "Parameters" is missing from "GetParametersCommand" output', () => {
    beforeEach(() => {
      ssmMock.on(GetParametersCommand).resolves({});
    });
    itRejects(handler, args);
  });

  describe('when "Parameters" are missing required values', () => {
    beforeEach(() => {
      ssmMock.on(GetParametersCommand).resolves({
        Parameters: [{ Name: 'dummyName' }],
      });
    });
    itRejects(handler, args);
  });
});
