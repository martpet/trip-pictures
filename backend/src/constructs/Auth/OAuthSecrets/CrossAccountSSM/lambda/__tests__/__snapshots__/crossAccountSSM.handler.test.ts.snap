// Vitest Snapshot v1

exports[`crossAccountSSM.handler > calls "getRoleCredentials" with correct args 1`] = `
[
  [
    "dummyRoleArn",
    "cross-account-ssm",
  ],
]
`;

exports[`crossAccountSSM.handler > resolves with a correct value 1`] = `
{
  "Data": {
    "values": [
      "dummyParameterValue1",
      "dummyParameterValue2",
    ],
  },
}
`;

exports[`crossAccountSSM.handler > sends "GetParametersCommand" from "SSMClient" with correct args 1`] = `
{
  "Names": [
    "dummyParameterName1",
    "dummyParameterName2",
  ],
}
`;

exports[`crossAccountSSM.handler > when "Parameters" are missing required values > rejects with a correct value 1`] = `[Error: Invalid SSM parameters]`;

exports[`crossAccountSSM.handler > when "Parameters" is missing from "GetParametersCommand" output > rejects with a correct value 1`] = `[Error: Invalid SSM parameters]`;

exports[`crossAccountSSM.handler > when "getParametersInput.Names" is missing > rejects with a correct value 1`] = `[Error: Missing input parameters names]`;
